### ISSUE TITLE
VBCSCompiler.exe crashes with a System.ArgumentException when building the default "Diagnostic with code fix" project

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#2495]
1. Install latest VSPro_1 daily drop (I have one from 29th April).

2. File -> New Project -> Download the .NET compiler platform SDK -> Download the SDK and restart VS.

3. Create a new "Diagnostic with code fix" C# project.

4. Build solution.

**Got:** VBCSCompiler.exe crashes with a `System.ArgumentException` with message: `The directory name ..\..\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C# is invalid.`

**Dump shared at:** \\manish-test\share_r\VBCSCompiler_Crash\VBCSCompiler.dmp

**Call stack:**
>   at System.IO.FileSystemWatcher..ctor(String path, String filter)
   at Microsoft.CodeAnalysis.CompilerServer.CompilerServerFileWatcher.AddPath(String path)
   at Microsoft.CodeAnalysis.CompilerServer.CSharpCompilerServer.RunCompiler(String clientDirectory, String[] args, String baseDirectory, String sdkDirectory, String libDirectory, IAnalyzerAssemblyLoader analyzerLoader, TextWriter output, CancellationToken cancellationToken, Boolean& utf8output)
   at Microsoft.CodeAnalysis.CompilerServer.CompilerRequestHandler.CSharpCompile(String currentDirectory, String libDirectory, String responseFileDirectory, String[] commandLineArguments, TextWriter output, CancellationToken cancellationToken, Boolean& utf8output)
   at Microsoft.CodeAnalysis.CompilerServer.CompilerRequestHandler.CSharpCompile(BuildRequest req, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CompilerServer.CompilerRequestHandler.HandleRequest(BuildRequest req, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CompilerServer.ServerDispatcher.Connection.<>c__DisplayClass6_0.<ServeBuildRequest>b__0()

**Comment by mavasani** [comment]
@tmeschter Is this a known/dupe issue?

**Comment by mavasani** [comment]
So, I went ahead and changed the project file for the default `Analyzer1 (Portable)` to use an absolute path instead of the relative path:

```
  <ItemGroup>
    <Analyzer Include="..\..\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.CSharp.Analyzers.dll" />
    <Analyzer Include="..\..\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\Microsoft.CodeAnalysis.Analyzers.dll" />
  </ItemGroup>
```

This works around the crash but the analyzers from these 2 CodeAnalysis analyzers are still unable to be loaded, I get the below diagnostics in output window:

Investigating further, I find that the analyzer assembly fails to load because it's dependency cannot be resolved:
`Could not load file or assembly 'System.Collections.Immutable, Version=1.1.33.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.`.

@tmeschter Does this mean that every project that needs an analyzer needs to explicitly add `System.Collections.Immutable` as an analyzer assembly OR does your proposed Consistency checker whitelist include this assembly? I am assuming it is the later, but still good to confirm.

**Comment by tmeschter** [comment]
@mavasani `System.Collections.Immutable` would be covered by a whitelist, but I haven't merged in the consistency checker yet.

Can you confirm that the correct version of System.Collections.Immutable is located next to the vbcscompiler.exe that is being run?

**Comment by mavasani** [comment]
@tmeschter Yes, version _1.1.34.0_ of System.Collections.Immutable is next to VBCSCompiler.exe. However, the analyzer project has nuget package references to older version (.33-beta) of the assembly.

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Microsoft.CodeAnalysis.Analyzers" version="1.0.0-rc1" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="Microsoft.CodeAnalysis.Common" version="1.0.0-rc1" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="Microsoft.CodeAnalysis.CSharp" version="1.0.0-rc1" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="Microsoft.CodeAnalysis.CSharp.Workspaces" version="1.0.0-rc1" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="Microsoft.CodeAnalysis.Workspaces.Common" version="1.0.0-rc1" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="Microsoft.Composition" version="1.0.27" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="NuGet.CommandLine" version="2.8.2" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="System.Collections.Immutable" version="1.1.33-beta" targetFramework="portable45-net45+win8" userInstalled="true" />
  <package id="System.Reflection.Metadata" version="1.0.18-beta" targetFramework="portable45-net45+win8" userInstalled="true" />
</packages>
```

Also, explicitly adding the following analyzer tag for `System.Collections.Immutable` didn't work for me as well:
```
<Analyzer Include="C:\Users\mavasani\Documents\visual studio 2015\Projects\Analyzer1\packages\System.Collections.Immutable.1.1.33-beta\lib\portable-net45+win8+wp8+wpa81\System.Collections.Immutable.dll" />
```

I see that this assembly doesn't get shadow copy'ed over to the temp path. Is it because there is a higher version of it lying next to VBCSCompiler.exe OR is it because we never attempted to load the assembly for this AnalyzerFileReference as it has no diagnostic analyzers?

**Comment by mavasani** [comment]
@tmeschter I have opened #2498 to track upgrading our template projects to newer nuget packages.

**Comment by mavasani** [comment]
@tmeschter I am hitting this crash even when avoiding use of the template project for writing an analyzer. As soon as I use "Add Analyzer" with full path to some assembly which has an analyzer, the IDE adds a relative path such as `<Analyzer Include="..\..\ClassLibrary1\ClassLibrary1\bin\Debug\ClassLibrary1.dll" />` to the project file and building the project this causes this crash.

**Comment by tmeschter** [comment]
@mavasani Right. The original problem is that we're not resolving the relative path to an analyzer into a full path, which causes `System.IO.FileSystemWatcher` to try and find the path relative to the current working directory. The working directory for vbcscompiler.exe will generally be the directory it resides in. So this code will still accidentally work when we build Roslyn (because we're picking up vbcscompiler.exe from a NuGet package that happens to be in the "correct" location relative to the analyzers) but will never work when using the vbcscompiler.exe that is installed with MSBuild.

I'm about to remove the use of `FileSystemWatcher` from vbcscompiler.exe, which will fix this problem.

**Comment by lgolding** [comment]
Verified as fixed. (If you follow the repro steps exactly, the compiler does not crash, but you do get the warnings related to System.Immutable mentioned above. However, that is fixed by #2498 .
